text/3519-arbitrary-self-types-v2.md

```diff
@@ -0,0 +1,485 @@
+/// There is no exclusivity; the underlying data may mutate, etc.
+/// (This is an abridged example: a real CppRef type would fully document invariants
+/// here.)
+pub struct CppRef<T: ?Sized> {
+    ptr: *const T,
+}
+
+impl<T: ?Sized> Receiver for CppRef<T> {
+    type Target = T;
+}
+
+// generated by bindings generator
+struct ConcreteCppType {
+    // ...
+}
+
+// all generated by bindings generator; mostly calls into C++
+// In this example these are not marked "unsafe" because we do not dereference
+// CppRef::ptr in Rust. However, arbitrary bad things may occur in foreign
+// languages so such functions are not fully "safe" either. Safety of FFI
+// is orthogonal to this RFC.
+impl ConcreteCppType {
+    fn some_cpp_method(self: CppRef<Self>) {}
+    fn get_int_field(self: &CppRef<Self>) -> u32 {}
+    fn get_more_complex_field(self: &CppRef<Self>) -> CppRef<FieldType> {}
+    fn equals(self: &CppRef<Self>) -> bool {}
+}
+
+// generated by bindings generator
+fn get_cpp_reference() -> CppRef<ConcreteCppType> {
+    // also calls into C++
+}
+
+fn main() {
+    // Rust code manipulating C++ objects via C++-semantics references
+    let cpp_obj_reference: CppRef<ConcreteCppType> = get_cpp_reference();
+    // cpp_obj_reference does not obey Rust reference semantics. Other
+    // "references" to the same data may exist in the Rust or C++ domain.
+    // But it can effectively be used as an opaque token to pass safely
+    // through Rust back into C++
+    let some_value: u32 = cpp_obj_reference.get_int_field();
+    let some_field = cpp_obj_reference.get_more_complex_field();
+    cpp_obj_reference.equals(&get_cpp_reference());
+}
+```
+
+(fuller example [here](https://github.com/google/autocxx/blob/main/src/reference_wrapper.rs#L117), with various [trait-based attempts](#not-do-it) to work around the lack of arbitrary self types.)
+
+Another case is when the existence of a reference is, itself, semantically important — for example, reference counting, or if relayout of a UI should occur each time a mutable reference ceases to exist. In these cases it's not OK to allow a regular Rust reference to exist, and yet sometimes we still want to be able to call methods on a reference-like thing.
+
+A third motivation is that taking smart pointer types as `self` parameters can enable functions to act on the smart pointer type, not just the underlying data. For example, taking `&Arc<T>` allows the functions to both clone the smart pointer (noting that the underlying `T` might not implement `Clone`) in addition to access the data inside the type, which is useful for some methods. Also, being able to change a method from accepting `&self` to `self: &Arc<Self>` can be done in a mostly frictionless way, whereas changing from `&self` to a static method accepting `&Arc<Self>` will always require some amount of refactoring. These options are currently open only to Rust's built-in smart pointer types, not to custom smart pointer types.
```

Comment:
```suggestion
A third motivation is that taking smart pointer types as `self` parameters can enable functions to act on the smart pointer type, not just the underlying data. For example, taking `&Arc<T>` allows the functions to both clone the smart pointer (noting that the underlying `T` might not implement `Clone`) in addition to access the data inside the type, which is useful for some methods; this also makes it ergonomic in more cases to make `Arc<SomeType>` explicit rather than having `SomeType` contain an `Arc` internally and have `Arc`-like `clone` semantics. Also, being able to change a method from accepting `&self` to `self: &Arc<Self>` can be done in a mostly frictionless way, whereas changing from `&self` to a static method accepting `&Arc<Self>` will always require some amount of refactoring. These options are currently open only to Rust's built-in smart pointer types, not to custom smart pointer types.
```

Adding another potential benefit: I've seen code that would have been clearer if written using an explicit `Arc<SomeType>`, but which has to wrap the `Arc` internally instead because `Arc<SomeType>` doesn't allow calling methods that can operate on the `Arc` (e.g. because the method needs to produce a `Weak` reference).

Reply:
Yep that's come up in some of the discussion, thanks.